### ISSUE TITLE
Creating a Perl wrapper exposes full source path and file name in the binary build

### ISSUE THREAD

**Issue by netx123** [swig/swig#2005]
Creating a wrapper for perl 5 using the -perl5 switch causes the resulting .dll file to expose the full source path and file name of the generated source file.

The issue is caused by the __\_\_FILE____ macro being used, looking at the generated file MYTEST_Perl_wrap.cxx shows the following:

 /* Install commands */
  for (i = 0; swig_commands[i].name; i++) {
    /* Casts only needed for Perl < 5.10. */
#ifdef __cplusplus
    newXS(const_cast<char*>(swig_commands[i].name), swig_commands[i].wrapper, const_cast<char*>(\_\_FILE__));
#else
    newXS((char*)swig_commands[i].name, swig_commands[i].wrapper, (char*)\_\_FILE__);
#endif
  }

**Comment by wsfulton** [comment]
Is there something wrong this that? newXS requires the filename. I suppose some other similar Perl api could be used, such as newXS_deffile, but it is not documented (https://perldoc.perl.org/5.6.0/perlapi) and I don't know what the downside is. Maybe @tlby has some thoughts on this.

**Comment by netx123** [comment]
Yes including \_\_FILE\_\_ in release builds leads to information disclosure where full file system paths of your source files and folders are unnecessarily included in release binaries.

It’s my understanding that the third parameter to newXS is purely used for Perl should it need to output debug/log messages, which would then provide the source of the file that triggered an error (***_Perl_wrap.cxx for example). So whilst there may be some merit in having this set for DEBUG builds, perhaps for RELEASE builds it would be more appropriate to leave the third parameter blank or have it set only to the **module** name.

**Comment by tlby** [comment]
As far as I know, this strategy was copied from [xsubpp](https://perldoc.perl.org/xsubpp).  I would hesitate to blaze our own trail in this area.  I'll try to see what accomodations [the library which provides this tool](https://metacpan.org/release/ExtUtils-ParseXS) might have.

**Comment by tlby** [comment]
I think `__FILE__` including the full path is an artifact of your build process.
```
$ cat try.c
#include <stdio.h>

int main(int ac, char *av[]) {
	puts(__FILE__);
	return 0;
}
$ gcc -o try try.c
$ ./try
try.c
$ gcc -o try ./try.c
$ ./try
./try.c
$ gcc -o try $PWD/try.c
$ ./try
/home/vagrant/tmp/try.c
$
```
So I think if full paths are not desired, the simplest strategy may be to avoid fully qualifying paths when the compiler is invoked.  [this thread](https://stackoverflow.com/questions/8487986/file-macro-shows-full-path) has some suggestions on how to get the compiler to suppress this information without restructuring how paths are resolved by your build scripts.

**Comment by ojwb** [comment]
This does seem like it's really an issue with build tools that use full paths.

But we could easily help out by using e.g. `SWIG_NEWXS_FILENAME` (better names likely exist) instead of `__FILE__` here and then stick this in `Lib/perl5/perlrun.swg`:

```
#ifndef SWIG_NEWXS_FILENAME
# define SWIG_NEWXS_FILENAME __FILE__
#endif
```

Then the user can just `#define SWIG_NEWXS_FILENAME ""` or whatever they want to override this.

**Comment by tlby** [comment]
Sure, but even if we had that indirection, if I had this problem myself I wouldn't bother with a -DSWIG_NEWXS_FILENAME.  I'd use solutions like [-ffile-prefix-map](https://gcc.gnu.org/onlinedocs/gcc/Overall-Options.html#index-ffile-prefix-map).

**Comment by ojwb** [comment]
`-ffile-prefix-map` is workable if you're only supporting GCC and/or clang, but isn't a portable solution.

And isn't redefining `__FILE__` undefined behaviour?

**Comment by tlby** [comment]
```
$ cat try.c
#line 1 "(hackers thwarted)"
#include <stdio.h>

int main(int ac, char *av[]) {
	puts(__FILE__);
	return 0;
}
$ cc -o try try.c
$ ./try
(hackers thwarted)
$
```
The Microsoft compiler claims to support the cpp `#line` directive, but I don't have a host to try it on.

I'm not saying I'm opposed to a `SWIG_NEWXS_FILENAME` patch, I'm just saying I wouldn't use the feature because it seems better solved by the tools that created the problem.  And I don't think that would be ideal for our friend above either, seems like the request is for SWIG to detect if it's a debug or a release build and handle things appropriately.  Closest chance I can think of to get there is hoping his release build feels the same way about `assert()` and hope `NDEBUG` is set.  🤷

**Comment by ojwb** [comment]
Ah yes, `#line` should provide a portable way to override `__FILE__` which is fairly clean, and has the big advantage of working with existing SWIG releases.

**Comment by netx123** [comment]
Thanks for looking into this, I prefer @ojwb idea of using SWIG_NEWXS_FILENAME , as this seems to be the most portable solution and provides a way to define the value to an empty value "", which would be ideal for me.

I'm also using MSVC, GCC and clang so SWIG_NEWXS_FILENAME would provide a consistent portable solution to remove the entire value across all compilers.

**Comment by ojwb** [comment]
There are two decent workarounds for this suggested above:

* Pass relative paths to SWIG
* Inject e.g. `#line 1 ""` near the start of the generated C/C++ wrapper file with:
```
%begin %{
#line 1 ""
%}
```

So given this has been opened for years with no patch proposed I'm going to close.

If someone is keen to have a better solution, they probably need to clarify the status of `newXS_deffile` with the Perl developers, or find a way to achieve this with documented Perl API features, and then propose a patch to change the code SWIG generates.
